doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+# The Life of a Connection
+
+This document describes, in broad strokes, how the Conduit proxy works with regard to
+application traffic. We assume that inbound and outbound routing are basically the same,
+modulo a few details that are called out in the proper context.
+
+## Accepting a connection
+
+When an application instance (app) calls a networked service, it resolves the DNS address
+of the service's name to an IP address. It initiates a connection to this address;
+however, because `conduit-proxy-init` has previously configured `iptables(8)`, the
+connection is established to the local `conduit-proxy`.
+
+When the proxy accepts a connection, it first determines the original destination address
+(i.e. the address bound by the client's DNS lookup) via the `SO_ORIGINAL_DST` socket
+option.
+
+It then determines whether there is policy associated with the remote peer's address
+and/or the original destination address. This policy may, for instance, pre-determine how
+the connection should be handled (i.e. with regard to TLS/HTTP/etc).
+
+### Protocol determination
+
+If protocol has not been pre-determined by policy, the proxy attempts to discover the
+protocol by reading data from the client. Note that this detection can break
+server-speaks-first protocols like MySQL and SMTP, necessitating the need to disable such
+detection by policy.
+
+In order to detect the connection's session protocol transparently (i.e. without policy),
+the proxy reads from the connection, buffering data until the proxy determines how to
+handle the connection.
+
+#### TLS
+
+If the connection begins with a TLS[[1]](#tls-q1) client handshake, data is read until
+until a full `ClientHello` message is buffered. The proxy determines whether it should
+terminate the TLS connection or treat the connection as an opaque TCP stream. The proxy
+may terminate TLS when a Conduit-specific extension[[2]](#tls-q2) exists in the client's
+hello, or it may terminate TLS if it is configured with a certificate for the
+client-requested name (i.e. for external/ingress).
+
+If the proxy elects to terminate TLS, the client's ALPN[[3]](#tls-q3) advertisement is
+used to determine which protocol should be used. If the ALPN advertisement does not
+include a proxy-supported protocol (i.e. `http` or `h2`), then the proxy handles the
+decrypted stream as opaque TCP[4](#tls-q4).
```

Comment:
The TLS ALPN extension works as follows:
1. The client either doesn't send the ALPN extension at all, or it sends a list of protocols in the ALPN extension in its TLS ClientHello message.
2. The server maybe sends the ALPN extension in its TLS ServerHello message.

If the client did not send an ALPN extension then it still is probably going to be HTTPS (HTTP over TLS) and we should be able to handle that.

The client could send {"h2, "something-we-don't-know"} and in that case we need to document what we do. I assume that we're going to pick some version of HTTP and ignore other protocols, whenever HTTP is in the list, and return that HTTP protocol's identifier in the ALPN extension in our ServerHello if we've decided to terminate TLS. Otherwise, I assume we're going to not send any ALPN extension in our ServerHello unless/until it's been explicitly configured.